feather/common/src/block.rs

```diff
@@ -0,0 +1,595 @@
+/// Blocks get replaced in-place when possible, while changing an adjacent block has a lower priority.
+fn place_block(
+    world: &mut World,
+    player_pos: Position,
+    chunk_entities: &ChunkEntities,
+    block: BlockId,
+    placement: &BlockPlacementEvent,
+    ecs: &Ecs,
+    light_level: u8,
+) -> Option<Vec<BlockChangeEvent>> {
+    let target1 = placement.location;
+    let target_block1 = world.block_at(target1.try_into().unwrap())?;
+    if target_block1.is_air() {
+        return None;
+    }
+    let mut block = block;
+    let player_dir_ordered = ordered_directions(player_pos.direction());
+    set_face(&mut block, &player_dir_ordered, placement);
+    rotate_8dir(&mut block, player_pos.yaw);
+    // Try placing the block by merging (=placing to the same position as the target block). Includes merging slabs, waterlogging and replacing blocks like grass.
+    let target = if slab_to_place(&mut block, target_block1, placement)
+        | waterlog(&mut block, target_block1)
+        | ((target_block1.kind() != block.kind()) & target_block1.is_replaceable())
+    {
+        merge_slab(&mut block, target_block1);
+        target1
+    } else {
+        // Otherwise place the block next to the target
+        let target2 = target1.adjacent(placement.face);
+        let target_block2 = world.block_at(target2.try_into().unwrap())?;
+        if merge_slab(&mut block, target_block2)
+            | waterlog(&mut block, target_block2)
+            | ((target_block2.kind() != block.kind()) & target_block2.is_replaceable())
+        {
+            target2
+        } else {
+            return None;
+        }
+    };
+    door_hinge(&mut block, target, &placement.cursor_position, world);
+    let place_top = match top_half(block) {
+        Some(_) => {
+            if !world
+                .block_at(target.up().try_into().unwrap())?
+                .is_replaceable()
+            {
+                // Short circuit if upper block is > 256
+                return None;
+            }
+            true
+        }
+        None => false,
+    };
+    let place_head = match bed_head(block) {
+        Some(_) => {
+            if !world
+                .adjacent_block_cardinal(target, block.facing_cardinal()?)?
+                .is_replaceable()
+            {
+                return None;
+            }
+            true
+        }
+        None => false,
+    };
+    // This works but there is a discrepancy between when the place block event is fired and getting the entity location.
+    // that makes it possible to place a block at the right exact moment and have the server believe it wasn't blocked.
+    if chunk_entities
+        .entities_in_chunk(target.chunk())
+        .iter()
+        .any(|_entity| {
+            let entity_position = ecs.get::<Position>(*_entity).unwrap();
+            let entity_kind = *ecs.get::<EntityKind>(*_entity).unwrap();
```

Comment:
Why is it `_entity` and not just `entity`?

Reply:
True.